### PR TITLE
Doc: Revive old-simulator-doc

### PR DESCRIPTION
**Pull request by Slayer95**
`simulator-doc.txt` lacks the detailed explanations from `old-simulator-doc.txt`.

This PR intends to bring the old documentation file up to date, so that it can be a reliable source, and in the future be merged into the main `simulator-doc.txt` file.

### CODE REVIEW

File: simulator-doc-extended.txt [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+	The duration counter of each effect is decremented automatically before
+	this event is fired (see duration for details).
+
+	examples: [volatile] Ghost-type Curse, [weather] Sandstorm
+
+
+=== DETAILED USAGE NOTES ===
+
+In Pokémon Showdown, as well as in Pokémon Lab (https://pokemonlab.com/) and
+DougJustDoug's framework, everything in Pokémon is an effect.
```
Comment (Slayer95) [inline simulator-doc-extended.txt]: What's this framework? I found this https://www.smogon.com/forums/threads/damage-calculator.65197/, but I can't be sure that's it.

**Comment by scheibo** [comment]
Ahah! Great minds think alike! This was on my [TODO list](https://pkmn.cc/TODO) as well! I'd like the merge them and update both.

File: simulator-doc-extended.txt (hunk 1 above)
Comment (Slayer95) [reply simulator-doc-extended.txt]: Note: I've already updated the content of this section.

File: simulator-doc-extended.txt [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+	NOTE: Not fired for abilities and items. For abilities and items, see
+	the switchIn event.
+
+	examples: [volatile] Encore, [side condition] Reflect,
+	          [volatile] Substitute
+
+restart(pokemon) [on status, on volatile]
+restart(side) [on side condition]
+restart() [on weather, on field effect]
```
Comment (Slayer95) [inline simulator-doc-extended.txt]: Note for self: singleEvent

File: simulator-doc-extended.txt [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+	Fired when trying to induce an effect that's already active.
+
+	Useful for effects that can have multiple "layers".
+
+	examples: [side condition] Toxic Spikes, [side condition] Pursuit,
+	          [volatile] Stockpile
+
+end(pokemon) [on status, on volatile, on item, on ability]
+end(side) [on side condition]
+end() [on weather, on field effect]
```
Comment (Slayer95) [inline simulator-doc-extended.txt]: Note for self: singleEvent

File: simulator-doc-extended.txt [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+	      before returning true.
+
+	Useful for giving messages that an effect has ended.
+
+	examples: [volatile] Encore, [side condition] Reflect,
+	          [volatile] Substitute
+
+duration(pokemon, source) [on status, on volatile]
+duration(side) [on side condition]
+duration() [on weather, on field effect]
```
Comment (Slayer95) [inline simulator-doc-extended.txt]: Note for self: "singleEvent", durationCallback

File: simulator-doc-extended.txt [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+
+movePriority(priority, user, target, move) [on user]
+foeMovePriority(priority, user, target, move) [on target]
+	Fired when determining a move's priority. Return the move's priority.
+
+	examples: [ability] Prankster
+
+start(pokemon, source) [on status, on volatile]
+start(side, source) [on side condition]
+start(placeholder, source) [on weather, on field effect]
```
Comment (Slayer95) [inline simulator-doc-extended.txt]: Note for self: singleEvent

File: simulator-doc-extended.txt [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+	Fired when a move hits (doesn't miss). Return true to prevent the move's
+	effects.
+
+	Also used for moves whose effects are too complicated to be represented
+	in any other way.
+
+	examples: [foe volatile] Protect, [move] Belly Drum, etc
+
+movePriority(priority, user, target, move) [on user]
+foeMovePriority(priority, user, target, move) [on target]
```
Comment (Slayer95) [inline simulator-doc-extended.txt]: Note for self: onModifyPriority

File: simulator-doc-extended.txt (hunk 1 above)
Comment (Zarel) [reply simulator-doc-extended.txt]: iirc he never ended up publishing it

**Comment by Zarel** [comment]
I assume this isn't done yet? I wrote these docs at some point but they always struck me as a confusing outdated mess; it's honestly easier to find events by reading the actual source code, for me.

**Comment by Zarel** [comment]
I can't imagine they'd ever be kept up to date even if we did reintroduce the docs, but I mean if you want to try, be my guest.

**Comment by Slayer95** [comment]
While it's certainly and humongous task to keep this fully updated, there are some things we can do.
1. We can comprehensively document non-move single events.
2. We can use this document to introduce the PS event system to new devs.

So what happens with move single events, and with the global events?
- ``singleEvent``s run on moves are currently expected to be documented in ``simulator-doc.txt`` (ofc the diagram is absolutely outdated after the latest spread move refactor).
- Global events' signatures are comprehensively documented in ``dev-tools/globals.ts``, as noted in this PR.

Regarding this PR, I expect comments about precision, style, and the like, for both goals 1 and 2. I am still missing some events to fully accomplish goal 1: ``onCopy`` for statuses, and all except ``onStart`` for items and abilities.

**Comment by Slayer95** [comment]
Rendered view so far: https://github.com/Zarel/Pokemon-Showdown/blob/7b3165ab996b4c8c88cf6f09d47fcdb37b219a64/SIM_EVENTS.md

**Comment by scheibo** [comment]
> I can't imagine they'd ever be kept up to date even if we did reintroduce the docs, but I mean if you want to try, be my guest.

If you're motivated, keeping docs (relatively) up to date isn't _that_ hard. Remembering during review to check if the changes would affect the docs is the best way, but this doesn't catch everything (though we could set up a [PR template](https://help.github.com/en/articles/creating-a-pull-request-template-for-your-repository) to remind contributors of this). At work, on Mondays I commit to looking over the previous week's worth of changes and checking to see if any would require my documentation to be updated and then updating the documentation all at once, meaning my project's docs are never more than ~1 week out of date.

I think the best way to avoid docs becoming out of date is to document the *right* things - ie/ high level stuff which doesn't change as frequently. While method signatures etc might subtly change, high level concepts like how the event system works etc are less likely to become obsolete as quickly (though admittedly, I've proposed reworking the event system, but I'd be more than happy to update the docs for that).

**Comment by Zarel** [comment]
If you want to like auto-render parts of this file from `dev-tools/globals.ts`, maybe even require comments, that might be nice for keeping things up to date.

It might also be hell. Like I said, an event registration syntax might be better than the current pure-declarative one.

**Comment by Slayer95** [comment]
> It might also be hell. Like I said in a different issue, an event registration syntax might be better than the current pure-declarative one.

While the future event registration API hasn't been designed yet, I think it would be in our best interest to keep (most of) the current event names and signatures

> If you want to like auto-render parts of this file from dev-tools/globals.ts, maybe even require comments, that might be nice for keeping things up to date.

I don't think adding a build step for the documentation is within scope, but that can be done in the future.

File: SIM_EVENTS.md [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+None | The event is run for its side effects. | `AfterBoost`, `Faint`, `SwitchIn`, `Update`
+[Cancellation token](#event-cancellation) | If the event is cancelled, the ongoing battle action is cancelled as well. | `DragOut`, `TryAddVolatile`, `TryHit`, `TryImmunity`
+Output | The return value of the event is passed to other functions. The event is very likely to support a [relay variable](#relay-variables). | `BasePower`, `Damage`, `ModifyAtk`, `ModifyCritRatio`
+
+### Event propagation
+
+Event propagation is the mechanism by which multiple effects have their event handlers
+called when a global event is run. They are called in order, from higher to lower priority,
+as per their listed Priority values. The priority should be specified with the key
+`&#96;${handlerName}Priority&#96;`, where `handlerName` is the full name of the associated event handler.
```
Comment (Slayer95) [inline SIM_EVENTS.md]: \`handlerNamePriority\`

**Comment by Acquiescence3** [comment]
I just updated this on another request. I didn't realize it had already been worked on. Sorry.

**Comment by AnnikaCodes** [comment]
No activity for 3 years—closing this.

**Comment by Slayer95** [comment]
There is some stuff that has gone outdated, but this is mostly still relevant, and above all, useful.

Reopen for visibility, please.

**Pull request closed without merging**